### PR TITLE
[5.5] Add Dynamic Enums to Eloquent

### PR DESCRIPTION
**Pull request by BrandonShar**
I've been working with Rails a lot lately and have gotten some good use out of the interesting way that Active Record handles enums. Instead of using the database enum field, Active Record creates an integer field on the table and uses an array on the model to determine the value for the user. 

I don't think this is a solution for all enum usages, but I think it makes for a great compromise between having the db speed and controlled options of an enum, but without being stuck needing to alter the database every time you need to modify the options. 

code could look like this:
```php
class MyModel extends Model
{
  protected $enums = [
    'color' => ['blue', 'yellow'],
  ];
}

$myModel = MyModel::first();
$myModel->color = 'yellow'; // stores 1 on the model and in the database row on save
echo $myModel->color; // yellow

$myModel->color = 'yllow';  // throws an InvalidEnumValueException on assignment
$myModel->color = null; // handles null fine.

// migrations
$table->enum('choices', ['yes', 'no', 'maybe']); // normal enum (current behavior)
$table->enum('choices'); // new "dynamic" enum field
```

### CODE REVIEW

**Comment by browner12** [comment]
we had a discussion about this on Slack a couple weeks ago.

I've recently switched out **all** of the enums from my database in favor of a string.

seems very dangerous to me to use the integers for storing in the database, as you have no real way to enforce that integrity. why not just store the string and be done with it?

one solution we had discussed was creating a `HasEnums` trait.  Remember we'll need a static way to access the possible values as well for when we are creating a new model.

**Comment by BrandonShar** [comment]
I know a few devs who always use strings instead of enums, but I've never totally bought into it. The advantages I think a technique like this has over enum strings are:
1. No risk of bad data. Typos will throw an exception instead of silently filling your database with a misspelled string
2. Having a "source of truth" for possible enum values. Another issue I disliked with using strings was having to refer to the database and sometimes run distinct queries just to see what possible values a column has. 
3. Slightly better data performance, integers take up less memory and query faster than strings

A static way to access access possible values makes a lot of sense. Maybe `$enums` should be a static property?

I should spend some more time in Larachat slack :smile:

**Comment by browner12** [comment]
1.  We might be misunderstanding each other here, but I'm saying do not use the **enum** type at all in your database. I'm saying use a simple 'string' type.  Enums are not consistently supported across all flavors of SQL. It can be expensive to change their acceptable values. Limited or negligible optimization. Truncates illegal values to empty strings.

2. Again, I don't want 'enum' types at all. I want the model to be the "source of truth". If your model contains all valid values, you can simply retrieve them from there.

3. I don't think the performance is consistently better, or even significantly better when it is. 

My big concern with using your integer based approach is easily broken data integrity.

Here is your original example:

```php
class MyModel extends Model
{
  protected $enums = [
    'color' => ['blue', 'yellow'],
  ];
}
```

If I save some 'yellow' items into the database they will be stored as `1`s.

Now what if I want to add the color green, but I want to keep them in alphabetical order.

```php
class MyModel extends Model
{
  protected $enums = [
    'color' => ['blue', 'green', 'yellow'],
  ];
}
```

All of a sudden, and very easily, my yellow items all became green.   If we simply store the values in the database as strings, our 'yellow' stays 'yellow'.

**Comment by BrandonShar** [comment]
No, I understood what you meant. My three points were all made pertaining to using a string field instead of an enum field.
1. With a string field,`$myModel->color = 'yllow'` is a fine statement and will silently add bad data to your database.
2. When you use a string field instead of an enum, it can be tricky to know what your "enum" options are. When you work with a table with 50k rows and maybe 4 potential options in your column, you're likely going to have to run `SELECT DISTINCT` to remember what the exact values are. 
3. querying on integer columns is faster than querying on varchar columns simply because integers are smaller (memory-wise). (because of the way enums are stored, they are also faster / smaller than string values) 

And yes, your example of rearranging the arrays makes it more complex, Active Record has the same difficulty. If you want to change the order after the fact, you would need to switch to an associative array and manually specify the keys. `[0 => 'blue', 2 => 'green', 1 => 'yellow']`. Not at all ideal, but functional if you really need to change the order of the values. I would aim to avoid rearranging the array after using it.

**Comment by BrandonShar** [comment]
I think a reasonable API for accessing the options would be using `__callStatic` to allow methods like 
```php
MyModel::colors(); // ['blue', 'yellow']
```
I'll wait to see the general consensus on adding this to core before I do more work on it though, otherwise I can turn it into a package.

Appreciate your feedback @browner12!

**Comment by browner12** [comment]
I'm still confused by your points.

1.  in your initial comment you say that an invalid value will throw an `InvalidEnumValueException`, so how would that go through to the database?

2. If we're adding the valid values to the model, why would you bother running a query at all? Just pull it off the model.  And a `SELECT DISTINCT` query only tells you what values are actually used. Let's say we have `ENUM('red', 'green', 'yellow')` but only 'red' and 'green' are used in the database, this query won't actually return the values you are looking for.

3. we could go back an forth on this, so I really have nothing to add to this point.

**Comment by sisve** [comment]
> With a string field,$myModel->color = 'yllow' is a fine statement and will silently add bad data to your database.

Who are you to say that the data is bad? Consider it an easy way to quickly add support for newly introduced enum values. `$model->color = 'blurple'`

> When you use a string field instead of an enum, it can be tricky to know what your "enum" options are. When you work with a table with 50k rows and maybe 4 potential options in your column, you're likely going to have to run SELECT DISTINCT to remember what the exact values are.

Not at all. If this is part of normal application functionality you have an index on it, and it would be an easy index operation to list all keys. In fact, this is a typical example when indexes are a good fit.

> querying on integer columns is faster than querying on varchar columns simply because integers are smaller (memory-wise). (because of the way enums are stored, they are also faster / smaller than string values)

Once again, indexes.

String values are the only solution that is easy maintainable; where I define maintainable as 1) easy to add new values and 2) easy to read existing values. It's also widely supported by all database engines.

**Comment by koenhoeijmakers** [comment]
I don't see the point in this, if you want enum to save values as "ids" you could have just made an inverse one to one or many (`BelongsTo`) relationship.

**Comment by koenhoeijmakers** [comment]
I don't really see the point in this.

1. Defining an inverse one to many (`BelongsTo`) relationship does exactly what you want, this also gives you the options to expand on this in the future.
2. You'll have ids in your db that refer to nothing, unless you know what is inside the model. (This makes your database more obscure, something you shouldn't ever want).

File: src/Illuminate/Database/Schema/Blueprint.php [hunk 1]
```diff
@@ -705,14 +705,18 @@ public function boolean($column)
 
     /**
      * Create a new enum column on the table.
+     * If they don't pass an array, assume they want
+     * a dynamic enum.
      *
-     * @param  string  $column
-     * @param  array   $allowed
+     * @param  string      $column
```
Comment (lucasmichot) [inline src/Illuminate/Database/Schema/Blueprint.php]: I don't think docblock variable names should be aligned...

File: src/Illuminate/Database/Schema/Blueprint.php [hunk 2]
```diff
@@ -705,14 +705,18 @@ public function boolean($column)
...
+     * @param  array|null  $allowed
      * @return \Illuminate\Support\Fluent
      */
-    public function enum($column, array $allowed)
+    public function enum($column, array $allowed = null)
```
Comment (lucasmichot) [inline src/Illuminate/Database/Schema/Blueprint.php]: Maybe don't enforce `array`, as we may want to use an object that implement `\ArrayAccess`

File: src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php [hunk 3]
```diff
@@ -195,6 +207,26 @@ protected function addCastAttributesToArray(array $attributes, array $mutatedAtt
     }
 
     /**
+     * Add dynamic enum attributes to the attributes array.
+     *
+     * @param  array $attributes
+     * @param  array $mutatedAttributes
+     * @return array
+     */
+    protected function addDynamicEnumAttributesToArray(array $attributes, array $mutatedAttributes)
```
Comment (lucasmichot) [inline src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: Don't force `array` here

File: src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php [hunk 4]
```diff
@@ -864,7 +904,7 @@ public function getAttributes()
     /**
      * Set the array of model attributes. No checking is done.
      *
-     * @param  array  $attributes
+     * @param  array $attributes
```
Comment (lucasmichot) [inline src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: Unrelated change

**Comment by BrandonShar** [comment]
Thanks for the comments, I'll try to address each of your points here:

`in your initial comment you say that an invalid value will throw an InvalidEnumValueException, so how would that go through to the database?`
I must have been unclear in response. I meant that with my dyanamic enums, assigning a non enum (like the obvious typo 'yllow') would throw an exception, and that is an advantage over using a regular string field, which would allow the misspelled value to be stored.

`If we're adding the valid values to the model, why would you bother running a query at all? Just pull it off the model....`
So you're storing potential values on the model as well? Are you using something similar to this to pick enum values? 

`we could go back an forth on this, so I really have nothing to add to this point.`
We actually can't go back on forth on this one. Integer columns are smaller and are therefore faster to query: this is a fact about SQL databases, not my opinion.

`Who are you to say that the data is bad? Consider it an easy way to quickly add support for newly introduced enum values.`
This is not the purpose of an ENUM. An enum explicitly exists to only allow data from a predefined set. If you want to be able to add data to your database on the fly, then don't consider an ENUM, just use a regular string column.

`If this is part of normal application functionality you have an index on it, and it would be an easy index operation to list all keys. In fact, this is a typical example when indexes are a good fit.`
Indexes on ENUMs aren't typically very beneficial. Remember that the point of an index is to make it significantly faster to drill down to the data you want so that any additional tablescans are done on a smaller data set. With an ENUM column, you have very few unique values, so eliminating data isn't as valuable. If you index an enum column with something like 3 values, at best you still have 1/3 of your table left to tablescan. Compare that to the cardinality of something like a foreign key and its a big difference. An index on a integer column is also still going to be faster than an index on a string column because of the speed of the comparisons. 

`Defining an inverse one to many (BelongsTo) relationship does exactly what you want, this also gives you the options to expand on this in the future.`
I think this is for a different use case. I think lookup tables are a great pattern, but aren't one to one replacements for an enum as they're a bit more cumbersome to use.

`You'll have ids in your db that refer to nothing, unless you know what is inside the model. (This makes your database more obscure, something you shouldn't ever want).`
Yes, this is definitely a drawback, but at the end of the day, my code is going to consume my data a lot more often than I am directly.

**Comment by tillkruss** [comment]
Cool, I'd love to see decent ENUM support in Eloquent.

**Comment by taylorotwell** [comment]
No plans to add this.

**Comment by BrandonShar** [comment]
Thanks for taking a look. I think I'll extract it into a package for anyone who's interested and comment back here when it's done.

**Comment by browner12** [comment]
here's one for people interested

https://gistlog.co/jhoff/b68dc2ac0e106a68488475e8430b38dc
